<a href="https://colab.research.google.com/github/RodolfoFigueroa/madi2023/blob/main/Unidad_1/01_Induccion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Antes de iniciar siempre es importante agregar al inicio las librerías que vamos a usar, y compilar la celda correspondiente.

In [2]:
import numpy as np

En esta sesión veremos algunos ejemplos de inducción y cómo este principio nos puede ayudar a ser más eficientes al resolver un problema.

# Ejemplos

## Ejemplo 1

Encuentra el valor de $$\Big\lfloor \frac{1}{2}\Big \rfloor + \Big\lfloor \frac{2}{2}\Big \rfloor + \cdots + \Big\lfloor \frac{10^9}{2}\Big \rfloor$$

**1.** Hagamos algunos casos pequeños para encontrar algún patrón.

In [35]:
s = 0
for n in range (1, 10):
    s += np.floor(n/2)
    print(f"n = {n}; Suma = {s}")

n = 1; Suma = 0.0
n = 2; Suma = 1.0
n = 3; Suma = 2.0
n = 4; Suma = 4.0
n = 5; Suma = 6.0
n = 6; Suma = 9.0
n = 7; Suma = 12.0
n = 8; Suma = 16.0
n = 9; Suma = 20.0


**2.** ¿Alguna conjetura? Probaremos por inducción que 

$$\Big\lfloor \frac{1}{2}\Big \rfloor + \Big\lfloor \frac{2}{2}\Big \rfloor + \cdots + \Big\lfloor \frac{n}{2}\Big \rfloor = \Big \lfloor \frac{n}{2} \Big \rfloor \Big \lfloor \frac{n+1}{2}\Big \rfloor$$

 - Caso base: Con los casos que hemos checado anteriormente es más que suficiente.
 - Hipótesis de inducción. La fórmula que queremos probar.
 - $n+1$: Supongamos que nuestra fórmula es cierta para $k$. Para $k+1$, sumamos $\lfloor(k+1)/2\rfloor$ de ambos lados.
 
$$\Big\lfloor \frac{1}{2}\Big \rfloor + \Big\lfloor \frac{2}{2}\Big \rfloor + \cdots + \Big\lfloor \frac{k}{2}\Big \rfloor + \Big\lfloor \frac{k+1}{2}\Big \rfloor = \Big \lfloor \frac{k}{2} \Big \rfloor \Big \lfloor \frac{k+1}{2}\Big \rfloor + \Big\lfloor \frac{k+1}{2}\Big \rfloor$$

 
$$= \Big \lfloor \frac{k+1}{2}\Big \rfloor (\Big \lfloor \frac{k}{2}\Big \rfloor + 1) $$

$$ = \Big \lfloor \frac{k+1}{2} \Big \rfloor \Big \lfloor \frac{k}{2}+1\Big \rfloor $$

$$ = \Big \lfloor \frac{k+1}{2} \Big \rfloor \Big \lfloor \frac{k+2}{2}\Big \rfloor $$



**3.** Concluimos calculando el valor que se nos ha pedido.

In [3]:
np.floor(1e9/2)*np.floor((1e9+1)/2)

2.5e+17

Adicionalmente, medimos cuánto tiempo tarda:

In [4]:
%%timeit
np.floor(1e9/2)*np.floor((1e9+1)/2)

2.26 µs ± 300 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


Comparemos el tiempo de ejecución con el tiempo que toma hacer la operación de forma directa. (Haremos la suma hasta $10^6$, pues tarda mucho para $10^9$).

In [14]:
%%timeit

s = 0
for i in range (1, 1000001):
    s += np.floor(i/2)

664 ms ± 5.36 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Esto nos muestra que fue de gran ayuda encontrar una fórmula cerrada, para reducir bastante el tiempo de ejecución.

## Ejemplo 2

Sea $f : \mathbb{N} \rightarrow \mathbb{R}$ una función tal que:

$$f(1) = 1001, \qquad f(1) + f(2) + \cdots + f(n) = n^2 f(n)$$ 

para todo entero $n > 1$. Encuentra el valor de $f(10007)$.

**1.** Por simplicidad, definimos:

$$g(n) \equiv \frac{f(n)}{1001}$$

Con lo cual tenemos:

$$g(1) = 1, \qquad g(1) + \cdots + g(n) = n^2 g(n)$$

Agrupando los $g(n)$ y factorizando:

$$(n^2 -1)g(n) = g(1) + g(2) + \cdots g(n-1)$$

Y como $((n-1)^2 - 1)g(n-1) = g(1) + g(2) + \cdots g(n-2)$, sustituyendo tenemos:

$$(n^2 -1)g(n) = ((n-1)^2 - 1)g(n-1) + g(n-1) = (n-1)^2g(n-1)$$

De esto obtenemos que: 

$$g(n) = \frac{n-1}{n + 1}g(n-1)$$ 

Ahora, hagamos los primeros casos:

In [32]:
g = 1
for n in range (2, 11):
    g = g*(n-1)/(n+1)
    print(f"n = {n} ; g(n) = {g}")

n = 2 ; g(n) = 0.3333333333333333
n = 3 ; g(n) = 0.16666666666666666
n = 4 ; g(n) = 0.1
n = 5 ; g(n) = 0.06666666666666667
n = 6 ; g(n) = 0.047619047619047616
n = 7 ; g(n) = 0.03571428571428571
n = 8 ; g(n) = 0.027777777777777776
n = 9 ; g(n) = 0.02222222222222222
n = 10 ; g(n) = 0.01818181818181818


**2.** Si recordamos la representación decimal de algunas fracciones, podemos ver que esto sigue un patrón. Para hacernos la vida más fácil, definimos: 

$$ h(n) = \frac{1}{g(n)}$$

$$= \frac{n+1}{n-1}h(n-1)$$

In [34]:
h = 1
for n in range (2, 11):
    h = h*(n+1)/(n-1)
    print(f"n = {n} ; h(n) = {h}")

n = 2 ; h(n) = 3.0
n = 3 ; h(n) = 6.0
n = 4 ; h(n) = 10.0
n = 5 ; h(n) = 15.0
n = 6 ; h(n) = 21.0
n = 7 ; h(n) = 28.0
n = 8 ; h(n) = 36.0
n = 9 ; h(n) = 45.0
n = 10 ; h(n) = 55.0


**3.** Podemos conjeturar que 

$$h(n) = \frac{n(n+1)}{2}$$

Probando por inducción:

* Caso base: 

$$h(2) = \frac{3}{1}\ h(1) = 3 $$

$$ \frac{2\cdot 3}{2} = 3$$

* H.I.: $h(n) = \frac{n(n+1)}{2}$

* Paso inductivo: Por definición:

$$h(n+1) = \frac{n+2}{n}h(n)$$

Por H.I.:

\begin{align}
&= \frac{n+2}{n}\frac{n(n+1)}{2}\\
&= \frac{(n+1)(n+2)}{2}
\end{align}

Finalmente, recordando que buscamos $f(10007)$, calculamos usando la fórmula que encontramos:

In [ ]:
print(2/(10007*10008*1001))

1.9950083682449443e-11


## Ejemplo 3

Definimos $F_n = 2^{2^n} + 1$, el $n$ -ésimo número de Fermat, para $n = 0, 1, \dots$. 

**Muestra que cualesquiera dos números de Fermat son primos relativos ($a$ y $b$ son primos relativos si su máximo común divisor es $1$).**

**1.** Casos pequeños. No podemos hacer muchos casos porque $2^{2^n}$ es muy grande, veamos qué pasa con los primeros $6$ números de Fermat:

In [38]:
F = []
for i in range(6):
    F.append(2**(2**i) + 1)
    print(F[i])

3
5
17
257
65537
4294967297


*Paso 2.* A primera vista no hay mucho de dónde agarrarnos. Intentando buscar un patrón notamos que:

* $3 = 5 -2$

* $3 \times 5 = 17 - 2$ 

* $3 \times 5 \times 17 = 255 = 257 - 2$

Entonces, podemos conjeturar que:

$$F_n - 2 = F_0\cdot F_1\cdot F_2 \cdots F_{n-1}$$

Para ver por qué nos sirve esto, consideremos $F_m$ y $F_n$, con $m>n$. Entonces, tenemos:

$$F_m - 2 = F_0\cdot F_1\cdot F_2 \cdots F_n\cdots F_{m-1}$$

Es decir:

$$2 = F_m - F_0\cdot F_1\cdot F_2 \cdots F_n\cdots F_{m-1}$$

Ahora, suponiendo que $d$ divide a $F_m$ y $F_n$, $d$ también debe de poder dividir la resta anterior; por lo tanto $d=1$ ó $d=2$. Sin embargo, es fácil ver que todos los números de Fermat son impares (ya que son una potencia de 2 mas 1), por lo tanto concluimos que $d=1$, y por lo tanto $F_m$ y $F_n$ son primos relativos.

---

Nuestra demostración anterior depende de la conjetura que hicimos:

$$F_n - 2 = F_0\cdot F_1\cdot F_2 \cdots F_{n-1}$$

Comprobemos que se vale para los primeros casos:

In [39]:
P = F[0]
for i in range (5):
    if (F[i+1] - P == 2):
        print("Se cumple")
    else:
        print("No se cumple :c")
  
    if i+1 < 5:
        P *= F[i+1]

Se cumple
Se cumple
Se cumple
Se cumple
Se cumple


*Paso 3*. Nuestros casos parecen soportarla. Entonces, probamos por inducción:

* Caso base: Los casos que ya vimos anteriormente.
* H.I.: $F_1 \cdot F_2 \cdot \dots \cdot F_k = F_{k+1}-2$. 
* Paso inductivo: Multiplicando ambos lados de la H.I. por $F_{k+1}$, tenemos:

$$F_1 \cdot F_2 \cdots F_{k} \cdot F_{k+1} = (F_{k+1} - 2)F_{k+1}$$

Recordando que $F_{n} = 2^{2^n}+1$, tenemos que:

\begin{align}
(F_{k+1} - 2)F_{k+1} &= (2^{2^{k+1}} - 1)(2^{2^{k+1}} + 1) \\
&= 2^{2^{k+2}} - 1 \\
&= F_{k+2} - 2
\end{align}

De donde concluimos el paso inductivo, y por lo tanto que se cumple $F_1 \cdot F_2 \cdots F_n = F_{n+1}-2$ para todo entero positivo $n$, lo cual a su vez prueba que todo par de números de Fermat son primos relativos

---

**Bonus:** Existe un número infinito de primos

**Demostración:** Hay un número infinito de números de Fermat, cada uno de los cuales es primo relativo en relación a todos los demás. Por lo tanto, cada $F_n$ tiene factores primos diferentes, lo cual prueba que hay un número infinito de primos.

## Ejemplo 4

Considera el conjunto $[k] = \{1, 2, \dots, k \}$. Definimos $S_k$ como el conjunto de todos los subconjuntos **no vacíos** de $[k]$ tales que no contienen parejas de enteros consecutivos, por ejemplo $\{1, 4\} \in S_k$, pero $\{2, 3, 7\} \notin S_k$. Adicionalmente, para un conjunto no vacío $B$, definimos $f$ como:

$$f(B) = \left(\prod_{x\in B}x\right)^2$$

Es decir, el cuadrado del producto de todos los elementos de $B$.

Finalmente, definimos $F_k$ como:

$$F_k = \sum_{B \in S_k}f(B)$$

$$ = \sum_{B \in S_k} \left(\prod_{x\in B}x\right)^2$$

Por ejemplo, suponiendo que tenemos $k=4$, el conjunto $S_4$ es:

$$S_4 = \left\{\{1,3\}, \{1,4\}, \{2,4\}, \{1\}, \{2\}, \{3\}, \{4\}\right\}$$

Entonces:

$$F_4 = (1\cdot 3)^2 + (1\cdot 4)^2 + (2\cdot 4)^2 + 1^2 + 2^2 + 3^2 + 4^4$$

$$ = 9 + 16 + 64 + 1 + 4 + 9 + 16$$

$$ = 119$$

**1.** En principio, podemos usar esta definición para escribir un programa para calcular los primeros valores, y así intentar encontrar un patrón. Sin embargo, esto es una tarea un tanto tediosa, ya que hay que calcular los subconjuntos que no contengan números consecutivos.

En vez de esto, reescribimos $F_k$ de la siguiente forma:

$$F_k = F_{k-1} + k^2\ F_{k-2} + k^2$$

Esto sale de considerar los conjuntos en $S_k$ que no contienen a $k$ (de lo que obtenemos el término $F_{k-1}$) seguido de los que sí lo hacen ($k^2 + k^2F_{k-2}$).

Usando esta fórmula, y dado que $F_0=0,\ F_1=1$, calculamos los primeros valores:

In [52]:
F = [0, 1]
for i in range(2, 7):
    F.append(F[i-1] + i**2 * F[i-2] + i**2)
    print(f"n = {i} ; Fn = {F[i]}")

n = 2 ; Fn = 5
n = 3 ; Fn = 23
n = 4 ; Fn = 119
n = 5 ; Fn = 719
n = 6 ; Fn = 5039


**2.** Notemos que al sumar $1$ a cada valor obtenemos un factorial, es decir:

$$F_k = (k+1)! - 1$$ 

Probemos por inducción que en efecto esto sucede:

* Caso base: Los casos que ya vimos.

* H.I.: $F_k = (k+1)! - 1$ para todo entero positivo $k \leq n$.

* Paso inductivo: Usando la ecuación derivada anteriormente:

$$F_{n+1} = F_n + (n+1)^2 F_{n-1} + (n+1)^2$$

Por H.I.:

\begin{align}
F_{n+1} &= [(n+1)! - 1] + [(n+1)^2\ (n! - 1)] + (n+1)^2 \\
&= (n+1)! + n!\ (n+1)^2 - 1 \\
&= (n+1)!\ (1 + n+1) - 1 \\
&= (n+2)! - 1
\end{align}

Con lo que concluimos que $F_{n+1} = (n+2)! - 1$.

Entonces la respuesta a nuestro problema inicial es $1001! - 1$, que es muy muy grande.

# Ejercicios

## Ejercicio 1

Considera la fórmula:

$$
f(n) = 5^n-1, \qquad n\geq 1, n\in\mathbb{N}
$$

* Escribe una función para encontrar el valor de $f(n)$ para $n$ arbitraria, que regrese un error si $n$ se sale del dominio.
* Utilizando tu función, encuentra para cuáles argumentos $n$, $f(n)$ es divisible por 3. Realiza una conjetura y demuéstrala por inducción.
* Haz lo mismo pero ahora para $f(n)$ divisible por 21.

## Ejercicio 2

Considera la suma:

$$
\frac{1}{1\cdot 2} + \frac{1}{2\cdot 3} + \frac{1}{3\cdot 4} + \cdots + \frac{1}{n\cdot (n+1)}, \qquad n\geq 1
$$

* Escribe una función que te permita calcular el valor de la suma para $n$ arbitraria, que regrese un error si $n$ se sale del dominio.
* Utiliza esta función y un bucle para intentar encontrar un patrón para $n$ pequeña.
* Conjetura una fórmula cerrada basada en tus resultados y demuéstrala por inducción.
* Utilizando esta fórmula, encuentra el valor de la suma para $n=123,456$